### PR TITLE
fix(moe): thread calibrated NVFP4 global scales through the unified MoE API

### LINKED ISSUE #3547: [Bug] Unified MoE API (#3093): trtllm_fp4_routed returns all-zeros for expert-parallel shards (local_expert_offset > 0)
**Affected:** unified MoE API, PR #3093 (`moe_api` branch), `flashinfer/fused_moe/runners.py`. SM100, NVFP4. `cute_dsl_nvfp4` is **not** affected.

### Summary
Driving `MoELayer` with an expert-parallel shard — `num_experts` (global) > `local_num_experts`, `local_expert_offset > 0`, weights prepared for the local shard, `selected_experts` passed as **global** ids (the contract `test_pack_inputs_applies_local_expert_offset` already encodes) — `trtllm_fp4_routed` produces **bit-exactly zero** output (`(out == 0).all()` is True). `cute_dsl_nvfp4` on the identical config is correct. EP is the standard large-MoE deployment mode, so this is a silent total wrong-answer on a common path.

### Repro / isolating diagnostic (H=I=512, T=128, top_k=4, SM100)
Cosine vs an independent fp32 reference, sweeping `(global, local, offset)`:
```
global=32  local=32 offset=0  : cute_dsl 0.9995   trtllm 0.9995   ok  (no EP)
global=64  local=32 offset=0  : cute_dsl 0.9995   trtllm 0.9995   ok  (global>local, offset 0)
global=64  local=32 offset=32 : cute_dsl 0.9995   trtllm 0.0000   BAD (all-zero)
global=128 local=32 offset=64 : cute_dsl 0.9995   trtllm 0.0000   BAD (all-zero)
```
The trigger is specifically **`local_expert_offset > 0`** — `global > local` at offset 0 is fine.

### Likely root cause — offset applied twice
`TrtllmFp4RoutedRunner` subtracts the offset in `pack_inputs` (`runners.py:257`: `ids = act.selected_experts - self._local_expert_offset`, producing local ids in `[0, local)`) **and also** forwards `local_expert_offset=self._local_expert_offset` to the kernel (`runners.py:305`). If the kernel subtracts it again, the ids fall outside `[0, local_num_experts)`, no expert matches, and the finalize writes nothing → all-zero. `cute_dsl_nvfp4` does **not** pre-subtract (it passes `selected_experts` as-is) and applies the offset once, so it is correct.

### Why CI missed it
`test_unified_moe.py::test_pack_inputs_applies_local_expert_offset` checks only the *packing* in isolation (offset ∈ {0, 32, 96}); there is no full EP-forward accuracy test, so the double-application in the kernel-call path is never exercised.

### Suggested fix
Apply the offset exactly once — either keep the `pack_inputs` subtraction and pass `local_expert_offset=0` to the kernel call, or stop pre-subtracting in `pack_inputs` and let the kernel own it (matching `cute_dsl`). Add a full EP-forward accuracy test.

(Found by a unified-API MoE fuzzer; file:line refs at PR #3093 head `89466b1a`.)

### LINKED ISSUE #3548: [Bug] Unified MoE API (#3093) only supports activation global_scale=1.0 — calibrated checkpoint scales silently inflate output (~2400x)
**Affected:** unified MoE API, PR #3093 (`moe_api` branch), `flashinfer/fused_moe/prepare.py` + `MoEActivationPack`. SM100, NVFP4.

### Summary
The MVP marshalling is wired **only for `global_scale = 1.0`**: `MoEActivationPack` has no field for the activation global scale, and `prepare_cute_dsl_nvfp4_weights` hardcodes `gs = torch.tensor([1.0])`, `fc2_input_scale = 1.0`, `alpha = ones`. As a result a caller cannot supply the **checkpoint-calibrated** scales real deployments use (vLLM/SGLang fold `w2_weight_scale * w2_input_scale` into `g2_alphas` / `output*_scale_scalar`).

It's also a **silent footgun**: if you quantize the activation with a real global scale — `a1_gs = (448*6)/max|x| ≈ 2.6e3`, i.e. what a calibrated checkpoint carries — the scale never reaches the kernel (no Pack field), and the output **inflates by roughly the activation global scale (~2400× for typical magnitudes)** with no error raised. Setting `a1_gs = 1.0` makes it correct.

### Why it matters
Not a correctness bug *at* `gs = 1.0` (block scales absorb a well-conditioned range), but it (a) blocks running a real ModelOpt/NVFP4 checkpoint through the unified API, and (b) is the exact surface of the recurring low-precision-MoE accuracy issues around small/calibrated intermediate scales.

### Repro
Drive `MoELayer` (NVFP4) from bf16 weights and build the activation pack with `a1_gs = (448*6)/x.abs().max()` instead of `1.0`. Output is ~2400× the reference (both backends still agree *with each other* — consistently wrong). With `a1_gs = 1.0` it matches the reference (≈ fp4 envelope).

### Suggested fix (matches the design doc's role-named QuantSpec, §4.3)
Add `hidden_states_scale_global` (and optionally per-expert) to `MoEActivationPack` and thread it into each runner's fc1 dequant; let `prepare_weights` accept calibrated weight + intermediate scales instead of hardcoding `1.0`/`ones`. Until then, document that the MVP is `global_scale = 1.0`-only so integrators don't pass calibrated scales and get silently-scaled output.

(Found by a unified-API MoE fuzzer; file:line refs at PR #3093 head `89466b1a`.)

### PR DESCRIPTION
**Pull request by waynehacking8**
## Description

**Stacked on #3591** (same API surface, the PRs are a series) — please review only the last commit (cb7a6a0); the first two are #3591.

The unified MoE API was wired only for `global_scale = 1.0`: `MoEActivationPack` had no field for the activation global scale, and both `prepare_*_weights` helpers hardcoded `gs = 1.0` / `ones` for every dequant scalar. Quantizing activations with a real calibrated scale (`a1_gs = (448·6)/amax ≈ 2.6e3`) silently inflated the output by ~that factor (#3548).

Changes:
- `MoEActivationPack` gains optional `hidden_states_scale_global` (`None` = 1.0, fully backward compatible).
- `prepare_trtllm_fp4_weights` / `prepare_cute_dsl_nvfp4_weights` accept calibrated `gemm1_scales_global` / `gemm2_scales_global` (per-expert, `None` → canonical `(448·6)/amax`) and `intermediate_scale_global`; weights are now actually quantized per expert with those scales, the weight-side dequant scalars use the canonical formulas, and the raw scales ride along in the view.
- Runners fold the activation scale in at pack time:
  - trtllm: `output1_scale_scalar` / `output1_scale_gate_scalar` ÷ `act_gs`, exactly matching the canonical gated-act formulas in `tests/moe/test_trtllm_gen_fused_moe.py` (`scale_c_fc1 = c_gs/(w1_gs·a1_gs)`, `scale_gate_fc1 = 1/(w1_gs·a1_gs)`, `scale_c_fc2` has no `a1_gs` term).
  - cute_dsl: `w1_alpha ÷ act_gs`. Formula evidence: the kernel computes `alpha·(A@B)` (`blockscaled_contiguous_gather_grouped_gemm_swiglu_fusion.py:332`) on a dequantized product carrying `a1_gs·w1_gs`; `fc2_input_scale` is the requant global scale itself (epilogue `q·SFC = acc·a2_gs`, `blackwell/...swiglu_fusion.py:2724-2789`, confirmed by `quant_dequant_fp4_reference` usage in `test_cute_dsl_fused_moe.py:180-183`); `w2_alpha = 1/(a2_gs·w2_gs)` (`...finalize_fusion.py:287`).
- New sm100-gated `TestUnifiedMoECalibratedScales`: end-to-end vs the bf16 reference with real `a1_gs`, per-expert weight scales, and a non-trivial intermediate scale.

## Related Issues

Fixes #3548. Stacked on #3591 (fixes #3547). Both found by the unified-API fuzzer.

### Pre-commit Checks

- [x] I have installed `pre-commit` by running `pip install pre-commit` (or used your preferred method).
- [x] I have installed the hooks with `pre-commit install`.
- [x] I have run the hooks manually with `pre-commit run --all-files` and fixed any reported issues.

## Tests

- [x] Tests have been added or updated as needed.
- [x] All tests are passing (`unittest`, etc.).

## Reviewer Notes

No SM100 hardware here (SM120 box), so beyond the full non-gated suite (98 passed) I hardware-verified what does run locally: per-expert quantization at `gs = 1` is **byte-identical** to the old flat path for both backends (packed weights *and* swizzled-sf bytes — the 128-row tile alignment argument holds), the trtllm scalars match the canonical formulas numerically, and a calibrated quant→dequant round-trip lands at rel-err ≈ 0.14 (fp4 envelope). The two new gated tests are the kernel-level proof and need an SM100 runner. Note the prepare helpers' default weight scale changed from `1.0` to the canonical per-expert `(448·6)/amax` (mirrors `quant_fp4_batches`); existing gated accuracy tests therefore now exercise calibrated weight scales — tolerances should hold (the reference uses the same scales) but that's worth one CI look. AI-assisted (Claude Code), reviewed line-by-line.

## Summary by CodeRabbit

* **New Features**
  * Added support for optional calibrated global scales for NVFP4 MoE weights and activations.
  * Added per-expert weight scaling and intermediate activation scaling for supported FP4 backends.
  * Added backward-compatible activation scale handling for pre-routed execution.

* **Bug Fixes**
  * Improved FP4 dequantization and activation scaling by incorporating provided calibration values.

* **Tests**
  * Added coverage verifying calibrated-scale results against reference MoE outputs across supported backends.

---
Rebased 2026-07-18 onto the routing-modes rework of `MoEActivationPack` and the test-helper move to `tests/moe/utils.py`: the new field is now keyword-only (matching the reworked pack's convention) and participates in its `__post_init__` same-device validation; the pack construction in the test uses the renamed `topk_ids` / `topk_weights`; `calibrated_activation_scale` ported into `tests/moe/utils.py::create_moe_tensors` (with the dequant reference scaled by `1/a1_gs`). File:line citations in the formula evidence above refer to the pre-rebase tree. Re-verified what runs on my SM120 box: the scale-resolver unit behavior (all-zero-expert guard, scalar broadcast, device check) and the full non-SM100 `test_unified_moe.py` suite (126 passed); the SM100-gated kernel tests collect but skip here.

### CODE REVIEW

File: flashinfer/fused_moe/runners.py [hunk 1]
```diff
@@ -236,16 +248,44 @@ def pack_inputs(
...
+                self._num_local_experts, device=a1_gs.device, dtype=torch.float32
+            )
+            # Views without precomputed scalars imply weight-side factors of
+            # 1.0 (weights quantized at global scale 1.0).
+            if output1_scale_scalar is None:
+                output1_scale_scalar = ones
+            if output1_scale_gate_scalar is None:
+                output1_scale_gate_scalar = ones
+            output1_scale_scalar = output1_scale_scalar * inv_a1
+            output1_scale_gate_scalar = output1_scale_gate_scalar * inv_a1
```
[the remark this answers is not in the document]
Comment (waynehacking8) [reply flashinfer/fused_moe/runners.py:426]: Fixed in 22a3689 — thanks. (For the amax guard I clamped to `torch.finfo(torch.float32).tiny` rather than a magic 1e-12.)

File: flashinfer/fused_moe/prepare.py [hunk 2]
```diff
@@ -35,6 +35,41 @@
...
+
+    ``None`` computes the calibrated scale ``(448 * 6) / amax`` per expert,
+    mirroring the canonical trtllm-gen quantize path
+    (``calculate_fp4_global_scale_factor`` / ``quant_fp4_batches`` in
+    ``tests/moe/test_trtllm_gen_fused_moe.py``).  A scalar broadcasts to all
+    local experts.
+    """
+    if scales is None:
+        amax = weights_bf16.float().abs().nan_to_num().amax(dim=(1, 2))
+        return (448.0 * 6.0) / amax
```
[the remark this answers is not in the document]
Comment (waynehacking8) [reply flashinfer/fused_moe/prepare.py]: Fixed in 22a3689 — thanks. (For the amax guard I clamped to `torch.finfo(torch.float32).tiny` rather than a magic 1e-12.)

File: flashinfer/fused_moe/runners.py [hunk 3]
```diff
@@ -94,8 +94,19 @@ def pack_inputs(
...
+        ``w1_alpha * (A @ B)`` and the dequantized NVFP4 product carries
+        ``act_gs * weight_gs``, so the combined alpha is the view's
+        weight-side ``w1_alpha`` (``1 / weight_gs``) divided by ``act_gs``.
+        ``fc2_input_scale`` / ``w2_alpha`` do not involve ``act_gs``.  ``None``
+        keeps the view tensors untouched (bit-identical to act_gs = 1.0).
         """
         v = weights.get_view(self.backend_key)
+        w1_alpha = v["w1_alpha"]
+        if act.hidden_states_scale_global is not None:
+            w1_alpha = w1_alpha / act.hidden_states_scale_global
```
[the remark this answers is not in the document]
Comment (waynehacking8) [reply flashinfer/fused_moe/runners.py]: Fixed in 22a3689 — thanks. (For the amax guard I clamped to `torch.finfo(torch.float32).tiny` rather than a magic 1e-12.)

**Comment by waynehacking8** [comment]
Rebased onto current main (`8fc7f079`, post-#3686) — now MERGEABLE. The functional fix is unchanged and still needed: on main, `MoEActivationPack` has no activation-global-scale field and `prepare.py` still hardcodes `global_scale=1.0` (gh #3548 remains live). Only merge adjustment: `create_moe_tensors` in `tests/moe/test_cute_dsl_fused_moe.py` now carries both main's new `gated` param and this PR's `calibrated_activation_scale` (orthogonal). Sits on top of the rebased #3591 as before. Verified locally on SM120 (CUDA 13): ruff clean; `test_unified_moe.py` + `test_cute_dsl_fused_moe.py` = 213 passed / 129 skipped (the `sm100_required` calibrated-scale accuracy tests need SM100 CI to execute).

**Comment by waynehacking8** [comment]
Rebased onto main again after #3591 landed (it touched the same MoE test files; branch rebuilt on 728658b2, now MERGEABLE). Functional fix unchanged; the fuzz test file from this branch is already upstream via #3591, so the diff is down to 5 files. Full tests/moe/test_unified_moe.py + test_cute_dsl_fused_moe.py pass on SM120 at the new head (213 passed, SM100-only cases skipped).

**Comment by waynehacking8** [comment]
Closing this stale implementation rather than rebasing it. Current main now has first-class `TrtllmFp4Config.prepare_activations`, while this branch does not integrate calibrated scale into that producer path, and the design document tracks #3548 under the standardized QuantSpec scale policy. The bug remains valid, so #3548 should stay open for that policy-level implementation.
